### PR TITLE
Remove pytest yield

### PR DESCRIPTION
**Pull request by nelzas**
## Description
handle issue: Pytest: yield tests are deprecated and must be removed. #2627

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

## References
#2627 

## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

File: skimage/transform/tests/test_radon_transform.py [hunk 1]
```diff
@@ -57,14 +57,22 @@ def check_radon_center(shape, circle):
...
-    for shape, circle in itertools.product(shapes, circles):
-        yield check_radon_center, shape, circle
-    rectangular_shapes = [(32, 16), (33, 17)]
-    for shape in rectangular_shapes:
-        yield check_radon_center, shape, False
+shapes_for_test_radon_center = [(16, 16), (17, 17)]
+circles_for_test_radon_center = [False, True]
+
+
+@pytest.mark.parametrize("shape", shapes_for_test_radon_center)
```
Comment (soupault) [inline skimage/transform/tests/test_radon_transform.py]: `itertools.product` here as well?

File: skimage/restoration/tests/test_unwrap.py [hunk 2]
```diff
@@ -109,10 +113,12 @@ def check_wrap_around(ndim, axis):
                         image_unwrap_wrap_around[index_last])
 
 
-def test_wrap_around():
-    for ndim in (2, 3):
-        for axis in range(ndim):
-            yield check_wrap_around, ndim, axis
+dim_axis = [(ndim, axis) for ndim in (2, 3) for axis in range(ndim)]
```
Comment (soupault) [inline skimage/restoration/tests/test_unwrap.py]: This is a pain to read :). Maybe reformulate somehow?

File: skimage/transform/tests/test_radon_transform.py [hunk 3]
```diff
@@ -105,12 +113,16 @@ def check_iradon_center(size, theta, circle):
...
-    thetas = [0, 90]
-    circles = [False, True]
-    for size, theta, circle in itertools.product(sizes, thetas, circles):
-        yield check_iradon_center, size, theta, circle
+sizes_for_test_iradon_center = [16, 17]
+thetas_for_test_iradon_center = [0, 90]
+circles_for_test_iradon_center = [False, True]
+
+
+@pytest.mark.parametrize("size", sizes_for_test_iradon_center)
```
Comment (soupault) [inline skimage/transform/tests/test_radon_transform.py]: `itertools.product`?

File: skimage/transform/tests/test_radon_transform.py [hunk 4]
```diff
@@ -261,9 +290,12 @@ def argmax_shape(a):
             argmax_shape(sinogram_circle_to_square))
 
 
-def test_sinogram_circle_to_square():
-    for size in (50, 51):
-        yield check_sinogram_circle_to_square, size
+sizes_for_test_sinogram_circle_to_square = (50, 51)
```
Comment (soupault) [inline skimage/transform/tests/test_radon_transform.py]: I think that short lists of parameters could be put directly into `pytest.mark.parametrize` (without specifying separate variable). Here and above

File: skimage/transform/tests/test_radon_transform.py [hunk 5]
```diff
@@ -294,13 +326,22 @@ def check_radon_iradon_circle(interpolation, shape, output_size):
...
+interpolations = ('nearest', 'linear')
+output_sizes = (None,
+                min(itertools.chain.from_iterable(shapes_radon_iradon_circle)),
+                max(itertools.chain.from_iterable(shapes_radon_iradon_circle)),
+                97)
+
+
+@pytest.mark.parametrize("shape", shapes_radon_iradon_circle)
+@pytest.mark.parametrize("interpolation", interpolations)
+@pytest.mark.parametrize("output_size", output_sizes)
```
Comment (soupault) [inline skimage/transform/tests/test_radon_transform.py]: One statement (via `product`)?

File: skimage/util/tests/test_dtype.py [hunk 6]
```diff
@@ -15,62 +15,65 @@
                np.float64: (-1.0, 1.0)}
 
 
+img_funcs_and_types = {img_as_int: np.int16,
```
Comment (soupault) [inline skimage/util/tests/test_dtype.py]: Maybe change to a tuple or use `zip` instead?

File: skimage/transform/tests/test_radon_transform.py [hunk 7]
```diff
@@ -294,13 +326,22 @@ def check_radon_iradon_circle(interpolation, shape, output_size):
...
 
 
-def test_radon_iradon_circle():
-    shape = (61, 79)
-    interpolations = ('nearest', 'linear')
-    output_sizes = (None, min(shape), max(shape), 97)
-    for interpolation, output_size in itertools.product(interpolations,
-                                                        output_sizes):
-        yield check_radon_iradon_circle, interpolation, shape, output_size
+# this might be overkill:
```
Comment (soupault) [inline skimage/transform/tests/test_radon_transform.py]: Indeed. Mad skills though :).
I'd prefer to put more code here, but stay as clear as possible.

File: skimage/restoration/tests/test_unwrap.py (hunk 2 above)
Comment (stefanv) [reply skimage/restoration/tests/test_unwrap.py]: Unfortunately, that's just how double comprehensions look in Python :/

**Comment by nelzas** [comment]
Hi,
Thanks for the comments. I'll try to delve into each of them in the next
week. I need to try to remember what led me to certain choices :-)

Nethanel

File: skimage/transform/tests/test_radon_transform.py (hunk 1 above)
Comment (nelzas) [reply skimage/transform/tests/test_radon_transform.py]: Agree, I will fix it.

File: skimage/transform/tests/test_radon_transform.py (hunk 3 above)
Comment (nelzas) [reply skimage/transform/tests/test_radon_transform.py]: Agree, I will fix it.

File: skimage/transform/tests/test_radon_transform.py (hunk 4 above)
Comment (nelzas) [reply skimage/transform/tests/test_radon_transform.py]: 1) OK, I will fix it.
2) I'm not sure where the "above" is.

File: skimage/transform/tests/test_radon_transform.py (hunk 7 above)
Comment (nelzas) [reply skimage/transform/tests/test_radon_transform.py]: My thinking was that sometime later one would want to test more than one shape and thus the complication. If you think it won't happen, I could go back to having one shape only as was before.

File: skimage/transform/tests/test_radon_transform.py (hunk 5 above)
Comment (nelzas) [reply skimage/transform/tests/test_radon_transform.py]: Sure, I'll fix it.

File: skimage/transform/tests/test_radon_transform.py (hunk 7 above)
Comment (stefanv) [reply skimage/transform/tests/test_radon_transform.py]: In general, write the simplest code that gets the job done.  We can always refactor later.

File: skimage/restoration/tests/test_unwrap.py (hunk 2 above)
Comment (nelzas) [reply skimage/restoration/tests/test_unwrap.py]: What about the example below? The logic is as before my changes but moved to a separate "builder" function:
```python
def generate_dim_and_axis():
    for ndim in (2, 3):
        for axis in range(ndim):
            yield ndim, axis
dim_axis = list(generate_dim_and_axis())
```
I personally prefer the double comprehension version, but open to change :-)

File: skimage/util/tests/test_dtype.py (hunk 6 above)
Comment (nelzas) [reply skimage/util/tests/test_dtype.py]: OK, I'll do it.

File: skimage/transform/tests/test_radon_transform.py (hunk 7 above)
Comment (nelzas) [reply skimage/transform/tests/test_radon_transform.py]: So what do you say about:
```python
shapes_radon_iradon_circle = ((61, 79), )
interpolations = ('nearest', 'linear')
output_sizes = (None,
                min(shapes_radon_iradon_circle[0]),
                max(shapes_radon_iradon_circle[0]),
                97)
```

**Comment by nelzas** [comment]
Hi,
So I pushed the changes I've already made. There are two comments that we still need to finalize.

File: skimage/restoration/tests/test_unwrap.py (hunk 2 above)
Comment (stefanv) [reply skimage/restoration/tests/test_unwrap.py]: I prefer the original too.

File: skimage/transform/tests/test_radon_transform.py (hunk 7 above)
Comment (stefanv) [reply skimage/transform/tests/test_radon_transform.py]: That looks fine to me.

**Comment by stefanv** [comment]
Good work, thanks for tackling this!

File: skimage/transform/tests/test_radon_transform.py (hunk 7 above)
Comment (nelzas) [reply skimage/transform/tests/test_radon_transform.py]: OK, I pushed this fix too.

**Comment by nelzas** [comment]
Thanks for letting me contribute something back :-)

File: skimage/transform/tests/test_radon_transform.py (hunk 4 above)
Comment (soupault) [reply skimage/transform/tests/test_radon_transform.py]: @nelzas I meant these lines - https://github.com/scikit-image/scikit-image/pull/2824/files#diff-fc1cd516a866b1cdcf5ed18a3d972af0R71, https://github.com/scikit-image/scikit-image/pull/2824/files#diff-fc1cd516a866b1cdcf5ed18a3d972af0R74

Comment (nelzas) [reply skimage/transform/tests/test_radon_transform.py]: No objection to fix the first one, I'm more reluctant about the second one as I find it less readable - a product of various lists without clear meaning:
```python
@pytest.mark.parametrize("size, theta, circle",
                         itertools.product([16, 17],
                                           [0, 90],
                                           [False, True])
```

Comment (soupault) [reply skimage/transform/tests/test_radon_transform.py]: @nelzas feel free to submit another PR with the refactoring of the current tests. :)
You've done a great job already, and I consider the original issue fixed, at this point.

**Comment by nelzas** [comment]
So le'ts agree to leave it like this for now. :-)
and thanks for the compliments.

Nethanel

**Pull request merged**